docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,196 @@
+expects, so that I can assume that the service will never proceed to install a
+cluster with a host that is running an incompatible version of the agent.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
+of the `AGENT_DOCKER_IMAGE` environment variable with the value of the
+`discovery_agent_version` stored in the `hosts` table.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, or in
+general any moving tag, as that renders the comparison useless. In the _SaaS_
+stage environment that variable is already using a tag, so that shouldn't be a
+problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the `preparing-for-installation`, `installing`,
+`installing-in-progress`, `error` or `cancel` states. If it is then the service
+will return the next steps as usual. If it isn't then the service will retrieve
+from the database the result of the `compatible-agent` validation. If the
+validation passed then the service will return the next steps as usual,
+otherwise it will replace all the usual steps the `upgrade-agent` step.
+
+When the agent receives the `agent-upgrade` step it will do nothing and return
+an exit code indicating that the agent needs to be upgraded. When receiving that
+exit code the next step runner will also exit, with the same code. The main
+program will react to that exit code starting the next step runner with the new
+image.
+
+The `upgrade-agent` step is also (similar to register) something we must not
+break in order to allow agent upgrade.
+
+### Risks and Mitigations
+
+The main risk is introducing interference with the state machine that controls
+the hosts. To reduce the chances of that the upgrade will be performed only when
+the host is not in the process of installing, i.e. not in one of the
+`preparing-for-installation`, `installing`, `installing-in-progress`, `error` or
```

Comment:
should we send agent upgrade on discovery state? I think we should write down which state upgrade will run with and not which states will be excluded